libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt

```diff
@@ -1588,6 +1530,157 @@ internal class MapRouteLine(
                 else -> y2
             }
         }
+
+        /**
+         * This will extract all of the road classes for the various sections of the route
+         * and determine the distance from the origin point to the starting point for each
+         * road class section and associate that data with a traffic congestion value.
+         *
+         * Each item returned represents a distance from the origin point, the road class
+         * beginning at that distance and the traffic congestion value for that section. The
+         * road class and congestion remain until the next item in the collections.
+         *
+         * The items returned are ordered from the point closest to the origin to the point
+         * farthest from the origin. In other words from the beginning of the route until the end.
+         *
+         * @param route the the calculations should be performed on.
+         */
+        fun getRouteLineTrafficExpressionData(route: DirectionsRoute):
+            List<RouteLineTrafficExpressionData> {
+                val distances = route.legs()?.mapNotNull {
+                    it.annotation()?.distance()
+                }?.flatten()
+                ifNonNull(distances) { distanceList ->
+                    val roadClassMap = mutableMapOf<Int, String>()
+                    route.legs()?.asSequence()
+                        ?.mapNotNull { it.steps() }
+                        ?.flatten()
+                        ?.mapNotNull { it.intersections() }
+                        ?.flatten()
+                        ?.filter {
+                            it.geometryIndex() != null
+                        }?.toList()?.forEach {
+                            roadClassMap[it.geometryIndex()!!] = it.mapboxStreetsV8()?.roadClass()
+                                ?: "intersection_without_class_fallback"
```

Comment:
I now discovered that the `geometryIndex` values are scoped per leg. This means that if we have a multi-leg route, the indices and class names will start to overwrite each other.

Reply:
https://github.com/mapbox/mapbox-navigation-android/pull/3672/commits/8ca2165d880d552b7cd2b04e87565ad973b87c43 adds the test case (currently failing).

Reply:
https://github.com/mapbox/mapbox-navigation-android/pull/3672/commits/40775c11927f31378c3bc40492de90eca05a5144 is a first fix attempt.